java/core/src/main/java/com/google/protobuf/DescriptorMessageInfoFactory.java

```diff
@@ -593,21 +616,88 @@ static String getFieldName(FieldDescriptor fd) {
     String name = (fd.getType() == FieldDescriptor.Type.GROUP)
                   ? fd.getMessageType().getName()
                   : fd.getName();
-    String suffix = specialFieldNames.contains(name) ? "__" : "_";
-    return snakeCaseToCamelCase(name) + suffix;
+
+    // convert to UpperCamelCase for comparison to the specialFieldNames
+    // (which are in UpperCamelCase)
+    String upperCamelCaseName = snakeCaseToUpperCamelCase(name);
+
+    String suffix = specialFieldNames.contains(upperCamelCaseName)
+                  // For field names that match the specialFieldNames,
+                  // append "__" to prevent field accessor method names from
+                  // clashing with other methods.
+                  ? "__"
+                  // For other field names, append "_" to prevent field names
```

Comment:
Should field names here be field method accessor names as well?

Reply:
No, the protoc compiler appends a single underscore specifically to the java field names to prevent clashes with java keywords.  The protoc compiler does not include this last undescore in accessor method names.